### ISSUE TITLE
Plans for DTLS 1.3

### ISSUE THREAD

**Issue by daenney** [pion/dtls#188]
I stumbled across https://datatracker.ietf.org/doc/draft-ietf-tls-dtls13/ today. Looks like it's on its way now.

I took a stroll through the spec, and https://tools.ietf.org/html/draft-ietf-tls-dtls13-34#section-12 has the changed compared to DTLS v1.2. I'm copying it here for good meassure:

   -  New handshake pattern, which leads to a shorter message exchange
   -  Only AEAD ciphers are supported.  Additional data calculation has
      been simplified.
   -  Removed support for weaker and older cryptographic algorithms
   -  HelloRetryRequest of TLS 1.3 used instead of HelloVerifyRequest
   -  More flexible ciphersuite negotiation
   -  New session resumption mechanism
   -  PSK authentication redefined
   -  New key derivation hierarchy utilizing a new key derivation
      construct
   -  Improved version negotiation
   -  Optimized record layer encoding and thereby its size
   -  Added CID functionality
   -  Sequence numbers are encrypted

Seems there's quite a bit of change. It might prove a fairly decent headache to support that.

**Comment by Sean-Der** [comment]
This is really exciting! Some random thoughts I had.

### We should join the IETF list and see if we can help.
Would be so exciting if Pion DTLS could make prototyping/building DTLS implementations faster.

### Put flights in separate packages
@at-wat split out all the code for flights into their own things. We can take advantage of this and have separate flights for 1.2 and 1.3 in their own packages.

A lot of stuff like the sequence numbers will be a pain to manage, but we will figure it out :)

**Comment by HRogge** [comment]
DTLS 1.3 is finally in RFC state...
https://www.rfc-editor.org/rfc/rfc9147.html

**Comment by eabase** [comment]
I would love to see some updates on this. 
`DTLS 1.3` is providing a some really essential features, regarding deep sleep and energy efficiency.

**Comment by Sean-Der** [comment]
I would love to help with this!

* Do any implementations of 1.3 exist (so we can interop test)
* Would any Pion users benefit from it/actually use it?

I am just hesitant to put work into this if no one ends up using it.

**Comment by HRogge** [comment]
We are working a lot with narrowband IP links at work (with tens or hundreds of kilobits/s channel capacity), so the more compact headers of DTLS1.3 and its zero-RTT reconnection would be a great way to secure UDP based communication (e.g. COAP).

Looking at OpenSSL ( https://github.com/openssl/openssl/issues/13900#issuecomment-1414440852 ) it mentions the WolfSSL implementation for interop testing DTLS 1.3

**Comment by beriberikix** [comment]
We're working on similar narrowband cellular problems like @HRogge (we should talk!) and have been tracking DTLS 1.3. That said, there's one feature, [Connection ID](https://datatracker.ietf.org/doc/rfc9146/), that was "backported" to 1.2 that provides significant power improvements that our customers and partners are asking for. @miguel91it has been evaluating what it would take to add it. Of course it's a separate issues from DTLS 1.3, but it's smaller and the thing we currently care most about, so I thought it would be worth sharing 🙂

**Comment by danielinux** [comment]
@eabase @HRogge 
Indeed [wolfSSL](github.com/wolfssl/wolfssl) supports DTLS 1.3 since release 5.4.0 (July 2022).

We would be interested in setting up some interoperability tests ourselves too, since we could not find any other implementation so far.

**Comment by HRogge** [comment]
Just forgot to add that [go-coap](https://github.com/plgd-dev/go-coap) refers to pion-dtls for TLS support... so everyone who wants to use "secure go-coap" will most likely also end using this project.

**Comment by eabase** [comment]
@danielinux 

> Indeed wolfSSL supports DTLS 1.3 since release 5.4.0 (July 2022).

No, that's incorrect. They support **TLS** `1.3`, but only **DTLS** `1.2`.
(Unless they're showing wrong info in the README.)

Corrected _wolfSSL_ link:
https://github.com/wolfssl/wolfssl

**Comment by danielinux** [comment]
@eabase Indeed the README is outdated. I've reported that to the team. Thanks for noticing that.

Confirming wolfSSL support both TLS 1.3 (since 2017) and DTLS1.3 (since July 2022).

**Comment by lakshya-chopra** [comment]
Are there any updates regarding this?

**Comment by daenney** [comment]
This is an open source project, and nobody is being paid to work on it. It may get supported if and when folks with the necessary knowledge and time are able to work on it.

Please refrain from pinging for status on this issue. It won't make things go any faster, nor will it result in us being able to make you a promise or give you a release timeline.

**Comment by HRogge** [comment]
Another issue here is that security specific library code is a bit more difficult to write than normal code... its easy to get wrong in subtile ways for people without experience.

Still, DTLS 1.3 is quite important for the future... maybe there is a way to get a funding pool for getting this job done?

**Comment by Sean-Der** [comment]
I am happy to take it on! I think we could make this happen by Jan 1st.

I need to release Pion WebRTC v4 first. After that I would say DTLS 1.3 is the most important thing.

If anyone has any advice/input/wants to get involved I would love to take the journey with a friend!

**Comment by theodorsm** [comment]
I have applied for funding through the NLnet foundation to dedicate my time to implement DTLS 1.3 and have been selected to enter the second round in the application process! Nothing is set in stone yet, and the next round is estimated to take six weeks.

The next step is to break down the main tasks and create a timeline. I will keep the community posted on progress and plans so we can all be aligned on the journey towards DTLS 1.3. With my workload otherwise, I aim to start towards the end of the year.

**Comment by theodorsm** [comment]
@Sean-Der thanks, glad to have you excited for this! It would be helpful to get feedback on the breakdown of tasks/features (when I have a draft ready) to ensure a well-scoped project for the final round of the application process. If the project gets the funding, having you or other maintainers available for PR reviews will be important to keep the project on track time-wise.

**Comment by Sean-Der** [comment]
If you want to start a doc/github issue/$x for breakdown I would love to help with that today! Will be useful no matter what.

I am here to approve/merge PRs right away. We should merge your stuff to master same day. I want to merge stuff as it is ready (and not a big PR).

This will have such a big positive impact. Faster connections for everyone and better privacy/security. Really hope NLNet works out :) 

If it doesn't would you be open to me raising money? I could reach out to companies using Pion and try to convince them to donate.

**Comment by theodorsm** [comment]
I will share a high-level plan here in this issue for discussion (as it seem to act as the discussion forum for DTLS 1.3 now), then we can break the plan down into more GitHub issues. It will take me a bit of time, so there isn't that much actionable for you @Sean-Der yet. I am happy to hear that you are motivated to merge stuff fast :)

I might be open to the idea of raising money, but let's circle back to that possibility only if NLnet declines the application.

**Comment by theodorsm** [comment]
I am very excited to share with the Pion community that I will receive funding from NLnet as part of the Next Generation Internet Commons Fund to implement DTLS 1.3. [https://nlnet.nl/project/PION-DTLS1.3/](https://nlnet.nl/project/PION-DTLS1.3/)

We still have to create the final plan together with NLnet, but so far I have: 

**Implement major changes in DTLS 1.3 defined in RFC 9147**
- Implement 1-RTT handshake.
- Implement 0-RTT handshake.
- Implement session resumption mechanism.
- Update key derivation hierarchy
- Implement ACK messages
- Implement encrypted sequence numbers 
- RFC 8446 Section 9.1: Add mandatory-to-implements cipher suites and signatures
- Add Post-Handshake Messages

**Implement new extensions from RFC 8446**
- pre_shared_key(41)
- early_data(42)
- supported_versions(43)
- cookie(44)
- psk_key_exchange_modes(45)
- certificate_authorities(47)
- oid_filters(48)
- post_handshake_auth(49)
- signature_algorithms_cert(50)
- key_share(51)

**Implement Encrypted Client Hello**

**E2E and unit testing of new functionality**

**Fuzz testing of parsing and perform protocol state fuzzing using [dtls-fuzzer](https://github.com/assist-project/dtls-fuzzer)**

**Interoperability testing**

**Adapt TLS 1.3 test cases from [TLS Anvil](https://github.com/tls-attacker/TLS-Anvil) to DTLS 1.3**

**Add DTLS 1.3 to pion WebRTC**

The deadline is August next year, but I hope to finish a bit before that. I will be working on this part-time starting in September, and full-time after the new year.

Please let me know if you are missing features in the plan above for the DTLS 1.3 implementation. I assume there might be some functionality I have overseen, but I will discover it when I first get into the implementation and dive into the RFCs. 

I will keep the community up-to-date going forward with the project.

**Comment by Sean-Der** [comment]
That is amazing news! When you are ready I would love to start merging small PRs :)

I want to make this as easy as possible.

**Comment by steely-glint** [comment]
This is great - lovely to see NLnet still funding good works after all these years!

One thing you might want to do is make sure that the new implementation is compatible with some moves that are being made in chrome/libwebrtc to speed up the multilayer handshake - specifically I'm told that the server hello might be embedded in the 'final' stun probe - withe the reply in the response (cutting out a full RTT). 

This means you may need to be able to handle wrapped DTLS packets as well as pure UDP ones - and unlike with TURN - not all the packets will come down the same route.

**Comment by JoTurk** [comment]
This is tracked in other issues now. And I think we'll have a working production ready MVP within two weeks.
